lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,73 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_namespace(o, name, file)
+    return if file.nil?
+
+    lc_file = file.downcase
+
+    file_namespace = @file_to_namespace[lc_file]
+    if file_namespace.nil?
+      return if @file_to_namespace.key?(lc_file)
+      file_namespace = @file_to_namespace[lc_file] = namespace_for_file(lc_file)
+      return if file_namespace.nil?
+    end
+
+    if !name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+  end
+
+  # The best option from https://stackoverflow.com/questions/1723182/a-regex-that-will-never-be-matched-by-anything
+  NEVER_MATCH = '\b'.freeze
+
+  def namespace_for_file(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+
+    return nil if initial_manifest(file)
+
+    path = path.each_filename.to_a
+
+    definition_dir_index = find_module_definition_dir(path)
+
+    # How can we get this result?
+    # If it is not an initial manifest, it must come from a module,
+    # and from the manifests dir there.  This may never get used...
+    return NEVER_MATCH if definition_dir_index.nil?
+
+    before_definition_dir = definition_dir_index == 0 ? 0 : definition_dir_index - 1
+    after_definition_dir = definition_dir_index + 1
+    names = path[after_definition_dir .. -2] # Directories inside module
+    names.unshift(path[before_definition_dir]) # Name of the module itself
+    # Do not include name of module init file at top level of module
+    if !(path.length == (after_definition_dir+1) && path[-1] == 'init.pp')
+      names.push(path[-1][0 .. -4]) # Remove .pp from filename
+    end
+
+    names.join("::").freeze
+  end
+
+  def initial_manifest(file)
+    manifest_setting = Puppet[:manifest]
+    full_path = Pathname.new(file).expand_path.to_s
+
+    #Does this ever happen outside of tests?
+    if manifest_setting.nil?
+      return full_path.end_with?('site.pp')
+    end
+
+    full_path == manifest_setting || full_path.start_with?(manifest_setting)
+  end
+
+  # Instead of 'manifests' our file could come from 'functions' or 'types'
```

Comment:
nit pick... "our file"? is it "ours" ? which file is that - the given path? (pet peeve: never talk about elements as "our")

How about saying "Returns module root directory index in path for files under 'manifests', 'functions' and 'types'"?